source/solvers/navier_stokes_vof_assemblers.cc

```diff
@@ -532,3 +528,386 @@ GLSNavierStokesVOFAssemblerSTF<dim>::assemble_rhs(
+  const unsigned int n_dofs     = scratch_data.n_dofs;
+  const double       h          = scratch_data.cell_size;
+
+  // Copy data elements
+  auto &strong_residual_vec = copy_data.strong_residual;
+  auto &strong_jacobian_vec = copy_data.strong_jacobian;
+  auto &local_matrix        = copy_data.local_matrix;
+
+  // Time steps and inverse time steps which is used for stabilization constant
+  std::vector<double> time_steps_vector =
+    this->simulation_control->get_time_steps_vector();
+  const double dt  = time_steps_vector[0];
+  const double sdt = 1. / dt;
+
+  // Phase values and limiters
+  std::vector<double> &phase_values = scratch_data.phase_values;
+  // std::vector<double> &phase_values_m1 =
+  // scratch_data.previous_phase_values[0];
+  // std::vector<Tensor<1, dim>> &phase_gradient_values =
+  // scratch_data.phase_gradient_values;
+
+  // Limit force application : not applied if cell density is below
+  // density_ratio of the maximum density (e.g. when one of the fluids is air)
+  const double density_ratio      = 2;
+  double       phase_force_cutoff = 0;
+
+  Assert(
+    scratch_data.properties_manager.density_is_constant(),
+    RequiresConstantDensity(
+      "GLSVansAssemblerDiFelice<dim>::calculate_particle_fluid_interactions"));
+
+  if (scratch_data.density_0[0] < scratch_data.density_1[0] &&
+      scratch_data.density_0[0] * density_ratio < scratch_data.density_1[0])
+    {
+      // gravity not will be applied for phase < phase_force_cutoff
+      phase_force_cutoff = 1e-6;
+    }
+  if (scratch_data.density_1[0] < scratch_data.density_0[0] &&
+      scratch_data.density_1[0] * density_ratio < scratch_data.density_0[0])
+    {
+      // gravity not will be applied for phase > phase_force_cutoff
+      phase_force_cutoff = 1 - 1e-6;
+    }
+
+  // Loop over the quadrature points
+  for (unsigned int q = 0; q < n_q_points; ++q)
+    {
+      const Tensor<1, dim> &velocity = scratch_data.velocity_values[q];
+      const Tensor<2, dim> &velocity_gradient =
+        scratch_data.velocity_gradients[q];
+      const Tensor<1, dim> &velocity_laplacian =
+        scratch_data.velocity_laplacians[q];
+      const Tensor<3, dim> &velocity_hessian =
+        scratch_data.velocity_hessians[q];
+      const Tensor<1, dim> &pressure_gradient =
+        scratch_data.pressure_gradients[q];
+
+      // Calculate shear rate (at each q)
+      const Tensor<2, dim> shear_rate =
+        velocity_gradient + transpose(velocity_gradient);
+
+      // Calculate the shear rate magnitude
+      double shear_rate_magnitude = calculate_shear_rate_magnitude(shear_rate);
+      // Set the shear rate magnitude to 1e-12 if it is too close to zero,
+      // since the viscosity gradient is undefined for shear_rate_magnitude = 0
+      shear_rate_magnitude =
+        shear_rate_magnitude > 1e-12 ? shear_rate_magnitude : 1e-12;
+
+      // Calculate viscosity gradient
+      const Tensor<1, dim> viscosity_gradient =
```

Comment:
Be careful, are you multiplying the viscosity gradient by the density? Don,t forget that the viscosity gradient in the VOF model is based on the dynamic viscosity